server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateEventsStep.java

```diff
@@ -0,0 +1,133 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.step;
+
+import com.google.common.base.Optional;
+import org.sonar.api.measures.CoreMetrics;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.core.measure.db.MeasureDto;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.event.Event;
+import org.sonar.server.computation.event.EventRepository;
+import org.sonar.server.computation.measure.MeasureRepository;
+
+public class QualityGateEventsStep implements ComputationStep {
+  private final TreeRootHolder treeRootHolder;
+  private final EventRepository eventRepository;
+  private final MeasureRepository measureRepository;
+
+  public QualityGateEventsStep(TreeRootHolder treeRootHolder, EventRepository eventRepository, MeasureRepository measureRepository) {
+    this.eventRepository = eventRepository;
+    this.measureRepository = measureRepository;
+    this.treeRootHolder = treeRootHolder;
+  }
+
+  @Override
+  public void execute() {
+    new DepthTraversalTypeAwareVisitor(Component.Type.PROJECT, DepthTraversalTypeAwareVisitor.Order.PRE_ORDER) {
+      @Override
+      public void visitProject(Component project) {
+        executeForProject(project);
+      }
+    }.visit(treeRootHolder.getRoot());
+  }
+
+  private void executeForProject(Component project) {
+    Optional<BatchReport.Measure> currentStatus = measureRepository.findCurrent(project, CoreMetrics.ALERT_STATUS);
+    if (!currentStatus.isPresent()) {
+      return;
+    }
+
+    Optional<MeasureDto> previousAlertStatus = measureRepository.findPrevious(project, CoreMetrics.ALERT_STATUS);
+
+    checkQualityGateStatusChange(project, currentStatus.get(), previousAlertStatus);
+  }
+
+  private void checkQualityGateStatusChange(Component project, BatchReport.Measure currentStatus, Optional<MeasureDto> pastStatus) {
+    String alertText = currentStatus.getAlertText();
+    MetricLevel alertLevel = MetricLevel.valueOf(currentStatus.getAlertStatus());
+    if (pastStatus.isPresent()) {
+      MetricLevel pastAlertLevel = MetricLevel.valueOf(pastStatus.get().getAlertStatus());
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Possible null pointer dereference in org.sonar.server.computation.step.QualityGateEventsStep.checkQualityGateStatusChange(Component, BatchReport$Measure, Optional) due to return value of called method [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)